File: day1.cpp
// Advent of Code 2022, day 1

#include <iostream>
#include <fstream>
#include <string>
#include <numeric>

using namespace std;

#define INPUT_FILE "input/1"
#define TOP_ELVES 3

int star_one();
int star_two();

int main() {
    cout << "star 1: " << star_one() << endl;
    cout << "star 2: " << star_two() << endl;
}

int star_one() {
    int max = 0;
    int sum = 0;

    ifstream input(INPUT_FILE);
    string line;
    while (getline(input, line)) {
        try {
            sum += stoi(line);
        } catch (invalid_argument e) {
            max = max > sum ? max : sum;
            sum = 0;
        }
    }

    return max;
}

int star_two() {
    int sum = 0;
    int top [TOP_ELVES] = { };
    int tmp;

    ifstream input(INPUT_FILE);
    string line;
    while (getline(input, line)) {
        try {
            sum += stoi(line);
        } catch (invalid_argument e) {
            for(int i = 0; i < TOP_ELVES; i++) {
                if (sum > top[i]) {
                    tmp = top[i];
                    top[i] = sum;
                    sum = tmp;
                }
            }

            sum = 0;
        }
    }

    return accumulate(top, top + TOP_ELVES, 0);
}

File: day1.rb
def star_one(input)
    max = 0
    sum = 0

    input.each_line do |line|
        if line == "\n" then
            max = [max, sum].max
            sum = 0
            next
        end
        sum += line.to_i
    end

    return max
end

def star_two(input)
    top = Array.new(3, 0)
    sum = 0

    input.each_line do |line|
        if line == "\n" then
            if sum > top.min then
                top[top.length - 1] = sum
                top.sort!.reverse!
            end
            sum = 0
            next
        end
        sum += line.to_i
    end


    return top.sum
end

input = File.read("input/1")
puts "star one: #{star_one(input)}"
puts "star two: #{star_two(input)}"
